ws/ShardManager.ts

```diff
@@ -0,0 +1,35 @@
+import DiscordWebsocket, { ConnectionProperties } from './DiscordWebsocket';
+
+interface BotGateway {
+  session_start_limit: {
+    max_concurrency: number;
+    remaining: number;
+    reset_after: number;
+    total: number;
+  };
+  shards: number;
+  url: string;
+}
+
+export default class ShardManager {
+  private shards: Map<number, DiscordWebsocket> = new Map();
+
+  constructor(private token: string, private intents: number, private connProps: Partial<ConnectionProperties> = {}) {
+    this.initialize();
+  }
+
+  private async getBotGateway() {
+    const botGateway: BotGateway = await (await fetch('https://discord.com/api/v8/gateway/bot')).json();
+    return botGateway;
+  }
+
+  private async initialize() {
+    const botGateway = await this.getBotGateway();
```

Comment:
@Khaaz do you want end user's to pass URL or no? I would prefer the former but discord states that ws URL is variable

Reply:
we want the user to be able to pass their own url. If none is passed we should be getting the gateway and change the gw url like that.

Reply:
not sure but we will need that to handle max concurency and how fast we want to identify each shard. For later though

Reply:
As mentioned, probably better if we allowed passing URL, and if there is none, then get the ws url then